mesa/datacollection.py

```diff
@@ -255,3 +255,25 @@ def get_table_dataframe(self, table_name):
         if table_name not in self.tables:
             raise Exception("No such table.")
         return pd.DataFrame(self.tables[table_name])
+
+
+class DataCollectorWithoutNone(DataCollector):
+    """
+    DataCollector where None values of the records are excluded
+    """
+
+    def _record_agents(self, model):
+        """Record agents data in a mapping of functions and agents."""
+        rep_funcs = self.agent_reporters.values()
+
+        def get_reports(agent):
+            _prefix = (agent.model.schedule.steps, agent.unique_id)
+            reports = (rep(agent) for rep in rep_funcs)
+            reports_without_none = tuple(r for r in reports if r is not None)
```

Comment:
Just too unpack this a little.. if the agent does not have the attribute, `None` is returned and and there will be no tuple and then no row in the `DataFrame` This prevents the issues we saw in Sugarscape with Traders where the Sugar and Spice information was retained as None and due to the number of sugar and spice agents this created memory issues. 

However, if the attribute is typed as a number then it will be collected as a NaN. But based on the dynamics of this should have less memory issues.

Reply:
> Just too unpack this a little.. if the agent does not have the attribute, None is returned and and there will be no tuple and then no row in the DataFrame This prevents the issues we saw in Sugarscape with Traders where the Sugar and Spice information was retained as None and due to the number of sugar and spice agents this created memory issues.

Yes, attributes which values are retrieved as `None` are excluded.

> However, if the attribute is typed as a number then it will be collected as a NaN. But based on the dynamics of this should have less memory issues.

This is not true, as long as you define the function to return `None` when appropriate. The NaN values only appear when the report get converted to a DataFrame. This is because a DF is constrained to be a table, so a N/A value for a number is NaN.